automatic axis orientation detection

linuxrider opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check. Remove checks that are not relevant and let us know if you need help with any of these.
-->
Checklist
* [x] Added an entry in `doc/news/`. <!-- Copy the TEMPLATE.rst to mybranch.rst, fill in the relevant sections, delete the others. -->
* [x] Added a test for this change.

<!--
Please add any other relevant info below:
-->

linuxrider commented:
@saraedum isort is putting an import above the license stuff. This looks strange to me.
Should we do something about it?

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -132,13 +134,26 @@
 
+class PageOrientation(Enum):
+    r"""Enum representing the page orientation to be able to to autodetect axis orientation."""
+    PORTRAIT = auto()
+    LANDSCAPE = auto()
```
```suggestion
    PORTRAIT = "portrait"
    LANDSCAPE = "landscape"
```

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -132,13 +134,26 @@
+
+class AxisOrientation(Enum):
+    r"""Enum representing axis orientation."""
+    HORIZONTAL = auto()
+    VERTICAL = auto()
```
```suggestion
    HORIZONTAL = "horizontal"
    VERTICAL = "vertical"
```

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -110,6 +110,8 @@
 
 """
+
+import itertools
```
```suggestion
```

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +570,35 @@ def axis_label(axis):
+        r"""
+        Return the reference points grouped by axis symbol.
+        """
+
```
```suggestion
         import itertools
```

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +569,35 @@ def axis_label(axis):
+        r"""
+        Return the reference points grouped by axis symbol.
+        """
+         import itertools
```
```suggestion
        import itertools
```

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -132,13 +133,26 @@
 #  along with svgdigitizer. If not, see <https://www.gnu.org/licenses/>.
 # ********************************************************************
 import logging
+from enum import Enum, auto
```
```suggestion
from enum import Enum
```

linuxrider commented:
So it seems that my locally installed black thinks differently than the one of the github actions.

saraedum commented:
> So it seems that my locally installed black thinks differently than the one of the github actions.

Are you using conda? If so, you can force the exact same version in your local installation.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
+            'j'
+
+        """
+
```
Does black really force these empty lines on us? That's weird.

linuxrider replied:
No it does not, checked it right now.

> > So it seems that my locally installed black thinks differently than the one of the github actions.
> 
> Are you using conda? If so, you can force the exact same version in your local installation.
> 
> We should probably set a version number in environment.yml to make it clear which version of black we are using.

Now, I have setup a dedicated environment from the provided yaml. It seems the work now.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -131,14 +132,29 @@
 
 
+class PageOrientation(Enum):
+    r"""Enum representing the page orientation to be able to to autodetect axis orientation."""
```
```suggestion
    r"""
    Page orientation of the source document.
    Used to auto-detect axis orientation.
    """
```

I don't really understand why this is needed. Do we allow inputs to be actually rotated here?

linuxrider replied:
Some whole page figures in old papers are in landscape orientation.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -131,14 +132,29 @@
+
+
+class AxisOrientation(Enum):
+    r"""Enum representing axis orientation."""
```
```suggestion
    r"""
    The orientation of a plot axis.
    """ 
```

Just trying to rewrite this so it does not exactly replicate what is written in the very line above it.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
         self.svg = svg
-        self.xlabel = xlabel or "x"
-        self.ylabel = ylabel or "y"
+        self.page_orientation = page_orientation
```
Why is this public? Should this start with a `_`?

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
+                self.page_orientation == PageOrientation.LANDSCAPE
+            ):
+                axis_orientations[axis] = AxisOrientation.HORIZONTAL
+            elif (delta_y > delta_x) ^ (
```
What if they are the same `==`?

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
+    @cache
+    def ylabel(self):
+        r"""
+        Return the label of the x axis.
```
The docstring and example always use the `x` axis.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
+
+        """
+
+        for axis, orientation in self.axis_orientations.items():
+            if orientation == AxisOrientation.HORIZONTAL:
+                return axis
```
```suggestion
        [axis] = [axis for axis in self.axis_orientations if self.axis_orientations[axis] == AxisOrientation.HORIZONTAL]
        return axis
```

I think this is shorter and has better error handling (reports an error if there is no such axis or more than one.)

linuxrider replied:
Very nice.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +572,35 @@ def axis_label(axis):
+
+        # group ref_points by axis
+        keyfunc = lambda x: str(x.label).split(":")[0][
+            :-1
```
Why do we drop the last character?

linuxrider replied:
This is removing the number which is not needed for grouping.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +572,35 @@ def axis_label(axis):
+        ref_points = sorted(ref_points, key=keyfunc)
+        # grouping by first letter of the label text
+        grouped_ref_points = {
+            quantity: list(g) for quantity, g in groupby(ref_points, keyfunc)
```
```suggestion
            variable: list(g) for variable, g in groupby(ref_points, keyfunc)
```

I think it is more generic to call this the variable of the plot.

linuxrider replied:
True.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +572,35 @@ def axis_label(axis):
+    @cache
+    def _grouped_ref_points(self):
+        r"""
+        Return the reference points grouped by axis symbol.
```
I had to read the code to understand what this does. Maybe you want to add an example here so it is clear what the returned dict looks like.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +572,35 @@ def axis_label(axis):
+            quantity: list(g) for quantity, g in groupby(ref_points, keyfunc)
+        }
+
+        if len(grouped_ref_points) > 2:
+            raise ValueError(
+                f"Reference points contain more than two axis names ({list(grouped_ref_points.keys())}). At the moment only 2D plots are supported."
+            )
```
```suggestion
        if len(grouped_ref_points) != 2:
            raise NotImplementedError(
                f"Currently, there must be exactly two axes since we only support 2D plots. However, we found the variables {list(grouped_ref_points.keys())} on the axes."
            )
```

Should be a NotImplementedError since you say that it's not implemented yet.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +572,35 @@ def axis_label(axis):
+        ref_points = [points[0] for points in self.labeled_paths["ref_point"]]
+        # without sorting only the last contiguous occurences are returned in the grouping dict
+        ref_points = sorted(ref_points, key=keyfunc)
+        # grouping by first letter of the label text
```
Why by the first letter of the label?

linuxrider replied:
This is not true anymore, it is a residue form the first approach.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -336,10 +572,35 @@ def axis_label(axis):
+            :-1
+        ]  # allow multi character axis label
+        ref_points = [points[0] for points in self.labeled_paths["ref_point"]]
+        # without sorting only the last contiguous occurences are returned in the grouping dict
```
I think that itertools is making things more complicated here. Maybe just:

```
def variable(point): return str(x.label).split(":")[0].strip()
variables = set(variable(point) for point in ref_points)
if len(variables) != 2:
    raise ...
grouped_ref_points = { v: [point for point in ref_points if variable(point) == v] for v in variables }
```

linuxrider replied:
I agree. itertools has been removed.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
+        for axis, points in self._grouped_ref_points.items():
+            # distinguish between 2 ref points
+            if len(points) == 2:
+                delta_x = abs(points[0].points[0][0] - points[1].points[0][0])
```
Why `points[0].points[0]`. Shouldn't this be `points[0].far`?

linuxrider replied:
True.

saraedum reviewed (changes requested):
It feels me that you are replicating a lot of code from `marked_points` here.

Wouldn't it be easier to just guess `xlabel` and `ylabel` and then look at the `transformation` matrix for `algorithm=mark-aligned` and pick the one that looks "less crazy"?

linuxrider commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +220,182 @@ class SVGPlot:
+        for axis, orientation in self.axis_orientations.items():
+            if orientation == AxisOrientation.HORIZONTAL:
+                return axis
+        return "x"
```
Removed.

linuxrider commented:
> It feels me that you are replicating a lot of code from `marked_points` here.
> 
> Wouldn't it be easier to just guess `xlabel` and `ylabel` and then look at the `transformation` matrix for `algorithm=mark-aligned` and pick the one that does less rotating?
> 
> It feels to me that such an approach would need very few lines of additional code.

How do I determine the amount of rotation?
The following is returned by `transformation` when (1) the right axis labels are passed and (2) when they are reversed.
```
array([[ 0.01,  0.  ,  0.  ],
           [ 0.  , -0.01,  1.  ],
           [ 0.  ,  0.  ,  1.  ]])

array([[ 0.  , -0.01,  1.  ],
           [ 0.01,  0.  ,  0.  ],
           [ 0.  ,  0.  ,  1.  ]])
```

linuxrider commented:
> > It feels me that you are replicating a lot of code from `marked_points` here.
> > Wouldn't it be easier to just guess `xlabel` and `ylabel` and then look at the `transformation` matrix for `algorithm=mark-aligned` and pick the one that does less rotating?
> > It feels to me that such an approach would need very few lines of additional code.
> 
> How do I determine the amount of rotation? The following is returned by `transformation` when (1) the right axis labels are passed and (2) when they are reversed.
> 
> ```
> array([[ 0.01,  0.  ,  0.  ],
>        [ 0.  , -0.01,  1.  ],
>        [ 0.  ,  0.  ,  1.  ]])
> 
> array([[ 0.  , -0.01,  1.  ],
>        [ 0.01,  0.  ,  0.  ],
>        [ 0.  ,  0.  ,  1.  ]])
> ```

I have no clue if this the correct method to determine the rotation, but I just compared the sums of the diagonals of the matrices.

linuxrider commented:
With the new approach, the determination of the x and y axis are postponed until it is really needed.
This is necessary since `transformation` needs `marked_points`.

linuxrider commented:
Since I changed 
```if (
                axis == self.xlabel
                and scalebar[0] < 0
                or axis == self.ylabel
                and scalebar[1] > 0
            ):
                scalebar = (-scalebar[0], -scalebar[1])
```
to
`scalebar = (abs(scalebar[0]), -abs(scalebar[1]))`
I added a test case for a x scale bar. I took the data from scaling_factor test case.
Point x2 was removed and a x scale bar was added with points on x1 and the formed x2 position.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -197,29 +208,163 @@ class SVGPlot:
+            {'E': <AxisOrientation.HORIZONTAL: 'horizontal'>, 'j': <AxisOrientation.VERTICAL: 'vertical'>}
+        """
+
+        from numpy import trace
```
It's a surprisingly reasonable approach to use the trace here. Since you told me that you are just guessing by looking at a few examples, let me try to explain why this is not a bad idea.

The affine transformation matrix is a 3×3 matrix. Essentially consisting of a 2×2 matrix in the top-left corner that describes a linear transformation and a 2×1 matrix in the top right that describes a translation. The rest of the matrix is most of the time just `0 0 1` in the bottom row.

(Not terribly import here, but your approach obviously fails if the bottom row is different in the two transformation matrices. Probably this does not happen here.)

So the question is: Looking at that 2×2 part, how can we estimate the amount of rotation that is happening?

Let's pretend for a moment that we just have a linear transformation and only look at the 2×2 matrix part and let's pretend that it's just a rotation by the angle θ. In that case, it's actually true that the trace is 1 + 2 cosθ. So, given the trace we can recover the absolute value of the angle. And more importantly, a small trace means rotation by a large angle, a small trace rotation by a large angle, at least when these angles are less that ±90°. (And a similar statement is true for angles close to 180° btw.)

The above explains why your approach works when the transformation is only a rotation.

Now what happens when the axes are differently scaled, i.e., when the transformation is not only a rotation matrix but a rotation matrix times a diagonal matrix. I think the approach still works because the scaling is compatible in the two transformations you compute.

I am not actually sure what happens when there is also shear. It might magically work again because the shear is somehow consistent in the two cases but I am not sure.

I think the safe approach here is to decompose the 2×2 matrix into a (rotation with flip) and a shear matrix. This is called a QR decomposition. Of this QR decomposition we can then compare the rotation angle with your approach.

---

So here's what we should do:

Grab the top left 2×2 matrix that describes the linear transformation of the affine transformation. QR decompose this matrix, this gives us a Q matrix which is orthogonal, i.e., a rotation, a reflection, or a combination of both. If the determinant of Q is 1, it's a rotation and so its trace measures the amount of rotation (the bigger the trace, the smaller the rotation.) If the determinant is -1, it's a reflection or a rotation+reflection. We can ignore the case that it's just a reflection since it's just a reflection and a rotation by 0°. So let's say it's a rotation+reflection. Then we undo the reflection by swapping the rows of the matrix Q and compute its trace and again, a large trace means little rotation, a small trace means big rotation.

For this to work, we do not even need the bottom row of the affine transformations to be the same `0 0 c`. It works even if they differ.

saraedum replied:
> Then we undo the reflection by swapping the rows of the matrix Q and compute its trace and again, a large trace means little rotation, a small trace means big rotation.

Sorry, that's nonsense. Such a swap corresponds to a flip with axis x=y.

linuxrider commented:
@saraedum I think it works now. Maybe you could another look if I implemented your suggestion correctly.

saraedum commented on svgdigitizer/electrochemistry/cv.py:
```diff
@@ -603,7 +607,9 @@ def _add_voltage_axis(self, df):
-        df["E"] = df["x"] * voltage.to(u.V).value  # pylint: disable=E1101
+        df["E"] = (
+            df[self.svgplot.xlabel] * voltage.to(u.V).value
+        )  # pylint: disable=E1101
```
For me, pylint does not complain about anything here.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -795,6 +1012,43 @@ def transformation(self):
+        """
+        return self._transformation(self.xlabel, self.ylabel, self._algorithm)
+
+    def _transformation(self, xlabel, ylabel, algorithm):
```
`algorithm` is only used in some places. In other places you use `self._algorithm`. I'll change this to be a static class method instead.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +215,172 @@ class SVGPlot:
+        from numpy.linalg import det, qr
+
+        axis_orientations = {}
+        case1 = self._transformation(
```
Since you are doing the same thing twice here, it should be possible to simplify this. Let me try.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +215,172 @@ class SVGPlot:
+        case2 = [sublist[:-1] for sublist in case2[:-1]]
+        q_2, _ = qr(case2, mode="complete")
+        if det(q_2) < 0:
+            case2 = case2[::-1]
```
This should be `q_2`.

saraedum commented on svgdigitizer/svgplot.py:
```diff
@@ -204,22 +215,172 @@ class SVGPlot:
+        if det(q_2) < 0:
+            case2 = case2[::-1]
+
+        if trace(case1) > trace(case2):
```
These should be the `Q` matrices.

saraedum commented:
@linuxrider, I fixed a bug in the algorithm (sorry, my algorithm had a bug) and simplified a few things (hopefully.)

Please have a look :)

linuxrider commented:
@saraedum I like your changes.
